### PR TITLE
Automatic Hint Cut for Class Records

### PR DESCRIPTION
**Pull request by SkySkimmer**
(https://github.com/coq/coq/issues/16684)

### CODE REVIEW

File: vernac/record.ml [hunk 1]
```diff
@@ -771,7 +771,27 @@ let interp_structure udecl kind ~template ~cumulative ~poly ~primitive_proj fini
...
+  let orl = fold (fun x y -> PathOr (x,y)) in
+  let starany = PathStar (PathAtom PathAny) in
+  let atom gr = PathAtom (PathHints [gr]) in
+  let projs = List.filter_map (fun p ->
+      Option.map (fun c -> atom (GlobRef.ConstRef c)) p.Structure.proj_body)
+      projs
+  in
+  if List.is_empty projs then ()
+  else
+    let entry = HintsCutEntry (seql [starany; orl projs; starany; atom (GlobRef.IndRef ind)]) in
```
Comment (JasonGross) [inline vernac/record.ml:791]: Shouldn't this be something like
```suggestion
    let entry = HintsCutEntry (seql [starany; orl projs; starany; atom (GlobRef.ConstructRef ind)]) in
```
?
We want the constructor, not the inductive itself.

Comment (SkySkimmer) [reply vernac/record.ml]: right

**Comment by SkySkimmer** [comment]
I expect no change but let's see
@coqbot bench

**Comment by SkySkimmer** [comment]
bench said
~~~
┌────────────────────────────┬───────────────────────────┬────────────────────────────────────────┬───────────────────────────────────────────┬─────────────────────────┐
│                            │       user time [s]       │               CPU cycles               │             CPU instructions              │  max resident mem [KB]  │
│                            │                           │                                        │                                           │                         │
│        package_name        │   NEW      OLD     PDIFF  │      NEW             OLD        PDIFF  │       NEW             OLD         PDIFF   │   NEW      OLD    PDIFF │
├────────────────────────────┼───────────────────────────┼────────────────────────────────────────┼───────────────────────────────────────────┼─────────────────────────┤
│              coq-fourcolor │ 1560.57  1564.99    -0.28 │  7128288205344  7143564017971    -0.21 │  12216895149858  12206108270041      0.09 │  737380   737228   0.02 │
│                   coq-core │  121.63   121.95    -0.26 │   492746238848   495577478941    -0.57 │    470890241118    471097328830     -0.04 │  286796   286816  -0.01 │
│               coq-coqprime │   46.92    46.94    -0.04 │   212697905321   212737046763    -0.02 │    310416964340    309128594657      0.42 │  764400   766280  -0.25 │
│     coq-mathcomp-odd-order │  435.00   434.61     0.09 │  1989937148616  1987822539738     0.11 │   3208248621931   3201109975072      0.22 │  959260   958228   0.11 │
│                coq-unimath │ 1230.43  1228.62     0.15 │  5615628850367  5604855387087     0.19 │   9912239334631   9915785980643     -0.04 │ 1602700  1600496   0.14 │
│     coq-mathcomp-ssreflect │   27.98    27.93     0.18 │   127368189192   127067158835     0.24 │    158509193400    157215264627      0.82 │  563964   566020  -0.36 │
│                 coq-stdlib │  488.59   487.66     0.19 │  2080613978393  2080044218865     0.03 │   1610965383250   1491647258738      8.00 │  697396   695900   0.21 │
│         coq-mathcomp-field │   96.10    95.77     0.34 │   439407186541   437778626185     0.37 │    693909493659    692490039357      0.20 │  647632   649488  -0.29 │
│      coq-engine-bench-lite │  175.28   174.58     0.40 │   760330458123   756580851115     0.50 │   1315461101437   1304618391658      0.83 │ 1063648  1020924   4.18 │
│      coq-mathcomp-fingroup │   23.48    23.38     0.43 │   107173331856   106522661485     0.61 │    152143238349    150624935309      1.01 │  497456   496972   0.10 │
│     coq-mathcomp-character │   70.25    69.78     0.67 │   320977935053   318259627674     0.85 │    469606790427    467615250954      0.43 │  740292   740216   0.01 │
│      coq-mathcomp-solvable │   84.86    84.22     0.76 │   387420261035   384028904655     0.88 │    568418913735    564681253717      0.66 │  660496   660908  -0.06 │
│       coq-mathcomp-algebra │   65.52    64.85     1.03 │   298856479217   295865199569     1.01 │    401709498563    396600068875      1.29 │  568880   570172  -0.23 │
│               coq-compcert │  312.80   305.64     2.34 │  1420630155219  1386888167536     2.43 │   2088518694606   2015617030059      3.62 │ 1124568  1125696  -0.10 │
│                coq-coqutil │   38.99    37.92     2.82 │   176135180639   170440772445     3.34 │    250103970464    236890199336      5.58 │  600244   600176   0.01 │
│                coq-bignums │   30.79    29.92     2.91 │   140998842567   136051215365     3.64 │    198199758228    185607362093      6.78 │  479360   479640  -0.06 │
│             coq-coquelicot │   38.86    37.39     3.93 │   174340398786   168084818228     3.72 │    232285425926    219603560513      5.77 │  753864   753896  -0.00 │
│   coq-metacoq-translations │   15.96    15.32     4.18 │    71643810456    69061508083     3.74 │    108701278719    104963850803      3.56 │  781560   782228  -0.09 │
│ coq-performance-tests-lite │  878.70   805.43     9.10 │  3987877493473  3649553582090     9.27 │   6866469873053   6100594689812     12.55 │ 2424312  2149740  12.77 │
│                  coq-verdi │   57.10    50.86    12.27 │   259606721772   230599004998    12.58 │    401678387795    336836430798     19.25 │  790228   791020  -0.10 │
│    coq-metacoq-safechecker │  257.10   225.19    14.17 │  1172649894461  1026852482323    14.20 │   1832660972820   1485926410211     23.33 │ 1451824  1445340   0.45 │
│              coq-fiat-core │   69.75    60.40    15.48 │   305583246312   262091096868    16.59 │    470577959454    362848407912     29.69 │  486812   487488  -0.14 │
│        coq-category-theory │  996.17   797.13    24.97 │  4554136250681  3640563078665    25.09 │   8041239109705   5930496668897     35.59 │  973712   913072   6.64 │
│       coq-metacoq-template │  177.63   141.46    25.57 │   802349204681   637127230762    25.93 │   1388677235978    993249157842     39.81 │ 1222320  1235492  -1.07 │
│                    coq-vst │ 1523.92   986.02    54.55 │  6961198410778  4494665381320    54.88 │  12854002040084   7005750009441     83.48 │ 2603912  2604268  -0.01 │
│        coq-metacoq-erasure │  355.17   219.40    61.88 │  1619283549965   994797602234    62.78 │   3036393184512   1522521265382     99.43 │ 1522472  1524888  -0.16 │
│               coq-rewriter │  625.62   378.38    65.34 │  2860126019516  1725231183462    65.78 │   5491741663113   2718143795622    102.04 │ 1233304  1231284   0.16 │
│                  coq-color │  401.02   239.13    67.70 │  1828165595104  1082310039011    68.91 │   3331601252458   1497035960713    122.55 │ 1182860  1184188  -0.11 │
│           coq-fiat-parsers │  687.10   361.01    90.33 │  3116454773674  1617718598218    92.65 │   6211141685085   2541053614283    144.43 │ 3156960  3157376  -0.01 │
│          coq-metacoq-pcuic │ 1319.23   596.36   121.21 │  6033281957340  2715847795544   122.15 │  11707801164396   3792984851951    208.67 │ 1955732  1952580   0.16 │
│                 coq-geocoq │ 2492.71   723.33   244.62 │ 11413276383817  3298731974576   245.99 │  23878759876885   5012961927176    376.34 │ 1010456   915132  10.42 │
│             coq-verdi-raft │ 4237.77   605.23   600.19 │ 19428238012651  2762737423230   603.22 │  41683951076999   4154990963908    903.23 │ 1230568  1218460   0.99 │
│                   coq-hott │ 1220.44   168.37   624.86 │  5589565035790   762710701482   632.86 │  11684174656766   1159620916885    907.59 │  622844   624776  -0.31 │
│          coq-iris-examples │ 9621.25   503.89  1809.39 │ 44112010880550  2290815983458  1825.60 │ 101662421169428   3317754572516   2964.19 │ 1169012  1186032  -1.44 │
│           coq-math-classes │ 5997.94    96.60  6109.05 │ 27509135390565   437845664039  6182.84 │  62641240730950    595075080521  10426.61 │  514908   513968   0.18 │
└────────────────────────────┴───────────────────────────┴────────────────────────────────────────┴───────────────────────────────────────────┴─────────────────────────┘
~~~
not that surprising tbh

**Comment by JasonGross** [comment]
What is going on here?  I thought we expected no change?  What's the usage pattern that this is preventing?

File: vernac/record.ml [hunk 2]
```diff
@@ -771,7 +771,27 @@ let interp_structure udecl kind ~template ~cumulative ~poly ~primitive_proj fini
...
+  let orl = fold (fun x y -> PathOr (x,y)) in
+  let starany = PathStar (PathAtom PathAny) in
+  let atom gr = PathAtom (PathHints [gr]) in
+  let projs = List.filter_map (fun p ->
+      Option.map (fun c -> atom (GlobRef.ConstRef c)) p.Structure.proj_body)
+      projs
+  in
+  if List.is_empty projs then ()
+  else
+    let entry = HintsCutEntry (seql [starany; orl projs; starany; atom (GlobRef.ConstructRef (ind,1))]) in
```
Comment (JasonGross) [inline vernac/record.ml:791]: Does the bench approve if we do
```suggestion
    let entry = HintsCutEntry (seql [starany; orl projs; atom (GlobRef.ConstructRef (ind,1))]) in
```
which should be good enough?

**Comment by SkySkimmer** [comment]
>What is going on here? I thought we expected no change? What's the usage pattern that this is preventing?

No idea but it feels unsurprising that changing typeclasses breaks existing code.

File: vernac/record.ml (hunk 2 above)
Comment (SkySkimmer) [reply vernac/record.ml]: let's see

**Comment by SkySkimmer** [comment]
@coqbot bench

**Comment by JasonGross** [comment]
I'd be interested in seeing the difference in typeclass debug logs from any of the "Top 25 slow downs" lines.  I don't suppose there's an easy way to get this though?

**Comment by SkySkimmer** [comment]
Should be easy to compile yourself?

**Pull request closed without merging**